// @ston-fi/sdk/dist/contracts/pTON/v1/PtonV1.js
import { address } from "@ton/ton";
import { JettonMinter } from "../../core/JettonMinter.js";
import { createJettonTransferMessage } from "../../../utils/createJettonTransferMessage.js";
import { pTON_VERSION } from "../constants.js";
const _PtonV1 = class _PtonV12 extends JettonMinter {
  constructor(address2 = _PtonV12.address, options) {
    super(address2, options);
  }
  async getTonTransferTxParams(provider, params) {
    const to = await this.getWalletAddress(provider, params.destinationAddress);
    const body = createJettonTransferMessage({
      queryId: params.queryId ?? 0,
      amount: params.tonAmount,
      destination: params.destinationAddress,
      forwardTonAmount: BigInt(params.forwardTonAmount ?? 0),
      forwardPayload: params.forwardPayload
    });
    const value = BigInt(params.tonAmount) + BigInt(params.forwardTonAmount ?? 0);
    return { to, value, body };
  }
  async sendTonTransfer(provider, via, params) {
    const txParams = await this.getTonTransferTxParams(provider, params);
    return via.send(txParams);
  }
};
_PtonV1.version = pTON_VERSION.v1;
_PtonV1.address = address(
  "EQCM3B12QK1e4yZSf8GtBRT0aLMNyEsBc_DhVfRRtOEffLez"
);
let PtonV1 = _PtonV1;
export {
  PtonV1
};
