// @ston-fi/sdk/dist/contracts/pTON/v1/PtonV1.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const ton = require("@ton/ton");
const JettonMinter = require("../../core/JettonMinter.cjs");
const createJettonTransferMessage = require("../../../utils/createJettonTransferMessage.cjs");
const constants = require("../constants.cjs");
const _PtonV1 = class _PtonV12 extends JettonMinter.JettonMinter {
  constructor(address2 = _PtonV12.address, options) {
    super(address2, options);
  }
  async getTonTransferTxParams(provider, params) {
    const to = await this.getWalletAddress(provider, params.destinationAddress);
    const body = createJettonTransferMessage.createJettonTransferMessage({
      queryId: params.queryId ?? 0,
      amount: params.tonAmount,
      destination: params.destinationAddress,
      forwardTonAmount: BigInt(params.forwardTonAmount ?? 0),
      forwardPayload: params.forwardPayload
    });
    const value = BigInt(params.tonAmount) + BigInt(params.forwardTonAmount ?? 0);
    return { to, value, body };
  }
  async sendTonTransfer(provider, via, params) {
    const txParams = await this.getTonTransferTxParams(provider, params);
    return via.send(txParams);
  }
};
_PtonV1.version = constants.pTON_VERSION.v1;
_PtonV1.address = ton.address(
  "EQCM3B12QK1e4yZSf8GtBRT0aLMNyEsBc_DhVfRRtOEffLez"
);
let PtonV1 = _PtonV1;
exports.PtonV1 = PtonV1;
